src/typing/env_js.ml

```diff
@@ -293,49 +342,93 @@ let already_bound_error =
    (since multiple declarations - sometimes but not always erroneous -
    may appear in an AST)
  *)
-let bind_entry cx name entry =
-  let scope = peek_scope () in
-  match Scope.get_entry name scope with
 
+let bind_entry cx name entry reason =
+  (* lex scopes can only hold let/const bindings
+   * var scopes can hold all bindings
+   * var bindings can't shadow lex bindings
+   * type entries are hoisted to var scope *)
+  let rec find_scope = function
+    | [] -> assert_false "empty scope list"
+    | scope::scopes -> Scope.(
+      match scope.kind, entry with
```

Comment:
Clearer and tighter to do nested matches on scope kind then entry kind here: 

    match scope.kind with
    | VarScope _ -> Some scope
    | LexScope -> Entry.(
      match entry.kind with
      | Let
      | Const -> Some scope
      | Type -> find_scope scopes
      | Var -> ...
    )

Also, I haven't worked it out completely but it doesn't seem right that both this function and the caller raise errors and need to do `get_entry` calls on the found scope. Don't know if it would be better to merge the two functions, or return an optional entry from here, or something else, but I think a cleaner separation into find-scope/do-stuff is possible. LMK if I'm talking nonsense and I'll take a crack at it. :)